sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala

```diff
@@ -226,11 +227,17 @@ object CSVRelation extends Logging {
           while (subIndex < safeRequiredIndices.length) {
             index = safeRequiredIndices(subIndex)
             val field = schemaFields(index)
-            rowArray(subIndex) = CSVTypeCast.castTo(
-              indexSafeTokens(index),
-              field.dataType,
-              field.nullable,
-              params.nullValue)
+            val value = indexSafeTokens(index)
+            val dataType = field.dataType
+            val isNull =
+              value == params.nullValue && field.nullable && !dataType.isInstanceOf[StringType]
+            val safeValue = dataType match {
```

Comment:
Is this logic unit tested? I see you removed the unit test related to `DecimalType`. I suggest you move this to a utility function and unit test it. Would also be good to make sure we have one test in CSVSuite as well.